### PR TITLE
Add medal emojis to top 3 scores

### PR DESCRIPTION
**Pull request by jgdhs27**
Add medal emojis (🥇, 🥈, 🥉) to the top 3 scores for a certain difficulty and shot.

image

In the future, I would like to add a player ranking by medal count, like royalflare had.

### CODE REVIEW

File: project/thscoreboard/replays/replays_to_json.py [hunk 1]
```diff
@@ -63,8 +77,21 @@ def _get_th17_additional_fields(self, shot: models.Shot) -> dict:
             "Goast": shot.GetSubshotName(),
         }
 
+    def _update_replay_count_by_category(self, difficulty: int, shot: models.Shot):
+        self.replay_count_by_category[(difficulty, shot)] += 1
+
     def convert_replays_to_serializable_list(
         self,
         replays: Iterable[models.Replay],
```
Comment (n-rook) [inline project/thscoreboard/replays/replays_to_json.py:85]: seems like we've introduced an undocumented requirement that replays be ordered (in reverse order) by score, no?

**Review by n-rook — CHANGES_REQUESTED** [review]

**Comment by jgdhs27** [comment]
Ok, I came up with another approach which adds the ranking inside the db query. I find this cleaner, I hope it addresses your concerns.

**Comment by jgdhs27** [comment]
Also, this makes it so that I could show the medals on the user page and front page without much difficulty, so I've done that.

File: project/thscoreboard/replays/replays_to_json.py [hunk 2]
```diff
@@ -65,6 +69,28 @@ def _get_th17_additional_fields(self, shot: models.Shot) -> dict:
...
-        replays: Iterable[models.Replay],
+        ranked_replays: Iterable[models.Replay],
     ) -> list[dict[str, any]]:
-        return [self._convert_replay_to_dict(replay) for replay in replays]
+        return [self._convert_replay_to_dict(replay) for replay in ranked_replays]
+
+
+def add_rank_annotation_to_replays(
+    replays: Manager[models.Replay],
+) -> Manager[models.Replay]:
```
Comment (n-rook) [inline project/thscoreboard/replays/replays_to_json.py:79]: Could you move this function over to models.py? We may want in places that aren't returning json, but it'd be strange to import "replays_to_json" in a place where we aren't converting replays to json. Plus, looking at the signature, I think you can define this as a method on a custom manager, if I remember Django right. Certainly calling replays.objects.some_query().annotate_with_rank() would be very convenient.

**Review by n-rook — CHANGES_REQUESTED** [review]
Nice, this definitely looks like a good way to do it. Just one more comment

File: project/thscoreboard/replays/replays_to_json.py (hunk 2 above)
Comment (jgdhs27) [reply project/thscoreboard/replays/replays_to_json.py]: Done. Although I think the current implementation isnt perfect - I can only call `.annotate_with_rank()` on a `ReplayQuerySet` object but not a `ReplayManager` object, which feels wrong.

Comment (jgdhs27) [reply project/thscoreboard/replays/replays_to_json.py]: disregard that, fixed that little issue too.

**Comment by jgdhs27** [comment]
Ok, another revision done. `annotate_with_rank` is now part of the `ReplayQuerySet` object, leading to much nicer code. Added tests to that. I also fixed a bug where TASes and irregular replays were counted in the rankings.

**Review by n-rook — APPROVED** [review]
Nice!

I'm _slightly_ worried about performance, but it's probably fine :)

**Pull request merged**